Sources/SourceKitLSP/SourceKitServer+Scheme.swift

```diff
@@ -0,0 +1,70 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LanguageServerProtocol
+import BuildServerProtocol
+
+extension SourceKitServer {
+
+  public func onSchemeChange(buildScheme: BuildScheme, workspace: Workspace) {
+    guard workspace.explicitIndexMode else {
+      return
+    }
+    collectTransitiveDependencies(
+      buildScheme: buildScheme,
+      workspace: workspace) { targets in
+        self.limitIndexVisibility(targets: targets, workspace: workspace)
+    }
+  }
+  
+  func collectTransitiveDependencies(
+    buildScheme: BuildScheme,
+    workspace: Workspace,
+    callback: @escaping ([BuildTargetIdentifier]) -> Void
+  ) {
+    workspace.buildSystemManager.buildTargets { targetsResponse in
+      guard case let .success(targets) = targetsResponse else {
+        callback([])
+        return
+      }
+      let targetMap = targets.reduce(into: [BuildTargetIdentifier: BuildTarget]()) {
+        $0[$1.id] = $1
+      }
+      var topLevelTargets = buildScheme.targets
+      var transitiveDeps = Set<BuildTargetIdentifier>()
+      while !topLevelTargets.isEmpty {
+        let targetID = topLevelTargets.removeLast()
+        if !transitiveDeps.contains(targetID), let target = targetMap[targetID] {
+          topLevelTargets.insert(contentsOf: target.dependencies, at: 0)
```

Comment:
Still O(n), why not just `topLevelTargets += target.dependencies`? Order makes no difference here, they're going in to a set anyway.

Reply:
Oh you mean the insert itself, From the previous comment I thought you meant to add a filter for duplicated deps targets

Reply:
Both required, the insert was the most painful part though.